evm/db/journal.py

```diff
@@ -0,0 +1,210 @@
+    of database keys and values.  The values are the "original" value of that
+    key at the time the checkpoint was created.
+    """
+    checkpoints = None
+
+    def __init__(self):
+        self.checkpoints = []
+        self.journal_data = {}
+
+    @property
+    def latest_id(self):
+        """
+        Returns the checkpoint_id of the latest checkpoint
+        """
+        return self.checkpoints[-1]
+
+    @property
+    def latest(self):
+        """
+        Returns the dictionary of db keys and values for the latest checkpoint.
+        """
+        return self.journal_data[self.latest_id]
+
+    @latest.setter
+    def latest(self, value):
+        """
+        Setter for updating the *latest* checkpoint.
+        """
+        self.journal_data[self.latest_id] = value
+
+    def add(self, key, value):
+        """
+        Adds the given key and value to the latest checkpoint.
+        """
+        if not self.checkpoints:
+            # If no checkpoints exist we don't need to track history.
+            return
+        elif key in self.latest:
+            # If the key is already in the latest checkpoint we should not
+            # overwrite it.
+            return
+        self.latest[key] = value
+
+    def create_checkpoint(self):
+        """
+        Creates a new checkpoint.
+        """
+        checkpoint_id = uuid.uuid4()
+        self.checkpoints.append(checkpoint_id)
+        self.journal_data[checkpoint_id] = {}
+        return checkpoint_id
+
+    def pop_checkpoint(self, checkpoint_id):
+        """
+        Returns all changes from the given checkpoint_id.
+        """
+        idx = self.checkpoints.index(checkpoint_id)
+
+        # update the checkpoint list
+        checkpoint_ids = self.checkpoints[idx:]
+        self.checkpoints = self.checkpoints[:idx]
+
+        # we pull all of the checkpoints *after* the checkpoint we are
+        # reverting to and collapse them to a single set of keys that need to
+        # be reverted.
+        revert_data = merge(*(
+            self.journal_data.pop(c_id)
+            for c_id
+            in reversed(checkpoint_ids)
+        ))
+
+        return dict(revert_data.items())
+
+    def commit_checkpoint(self, checkpoint_id):
+        """
+        Collapses all changes for the givent checkpoint into the previous
+        checkpoint if it exists.
+        """
+        changes_to_merge = self.pop_checkpoint(checkpoint_id)
+        if self.checkpoints:
+            # we only have to merge the changes into the latest checkpoint if
+            # there is one.
+            self.latest = merge(
+                changes_to_merge,
+                self.latest,
+            )
+
+    def __contains__(self, value):
+        return value in self.journal_data
+
+
+class JournalDB(BaseDB):
+    """
+    A wrapper around the basic DB objects that keeps a journal of all changes.
+    Snapshots are simply indices into the list of journal entries.  Reversion
```

Comment:
This sentence is no longer accurate as now the journal entries are kept in a dict with uuids as keys